### PR TITLE
docs: logging configuration guide

### LINKED ISSUE #3398: Document various useful logging.yaml config examples
Migrated from GitLab: https://gitlab.com/meltano/meltano/-/issues/3478

Originally created by @pandemicsyn on 2022-05-11 16:45:51

---

Would be nice to document various logging.yaml config examples. Couple of ideas:

1. A config examples for local dev
2. A general go to prod config 
3. A go to prod config where the console is consumed by a machine
4. A go to prod config where a log file is consumed by a machine
   - a variation showing automatically rotated logs for long lived deploys 
5. explicit config example to turn off colored console output
6. explicit config example that uses a custom log line format
7. explicit config examples showing filtering

/cc @vischous any other explicit examples that you think would useful ?

### PR DESCRIPTION
**Pull request by pandemicsyn**
Includes:

- A primer on how logging works
- The different parts of the logging yaml (with lots of links for folks to go explore and RTFM the python docs)
- A example local config
- Configs for GCP and Datadog (including a very basic Datadog Agent example config)
- A list of some common fields and their description
- A quick `jq` tip to get folks 🤔  about how all they can slice and dice locally

closes: https://github.com/meltano/meltano/issues/3398

### CODE REVIEW

**Comment by pandemicsyn** [comment]
@tayloramurphy we've chatted about doing this for users a bunch and I really wanted to finally get this documented, so did this outside of an official iteration - hope thats cool 🙏

**Comment by pandemicsyn** [comment]
@pnadolny13 are you using cloudwatch to capture logs for squared or anything? 

/cc @aaronsteers you mentioned you've used cloudwatch a lot in the past - any specific recipe we should include for CloudWatch alongside the Datadog and GCP ones or does the generic version cover it ?

File: docs/src/_tutorials/logging.md [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+Logging in meltano can be controlled via a standard yaml formatted [python logging dict config file](https://docs.python.org/3/library/logging.config.html#configuration-dictionary-schema).
+
+By default, meltano will look for this in a `logging.yaml` file in the project root. However, you can override this by
+setting the [environment variable](/guide/configuration#configuring-settings) `MELTANO_CLI_LOG_CONFIG` or by using the
+`meltano` CLI option `--log-config`. e.g. `meltano --log-config=my-prod-logging.yaml ...`.
+
+A logging.yaml contains a few key sections that you should be aware of.
+
+- `formatters` - This section contains the formatters that are used by the handlers. This controls the output format of the log messages (e.g. json).
+- `handlers` - This section contains the handlers which are used the loggers. This controls the output destination of the log messages (e.g. the console).
```
Comment (tayloramurphy) [inline docs/src/_tutorials/logging.md:18]:
```suggestion
- `handlers` - This section contains the handlers which are used by the loggers. This controls the output destination of the log messages (e.g. the console).
```

File: docs/src/_tutorials/logging.md [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+  my_warn_file_handler: # log everything WARNING and above to automatically rotating log file in key_value format
+    class: logging.handlers.RotatingFileHandler
+    level: WARN
+    formatter: key_value
+    filename: /tmp/meltano_warn.log
+    maxBytes: 10485760
+    backupCount: 20
+    encoding: utf8
+
+root:
```
Comment (tayloramurphy) [inline docs/src/_tutorials/logging.md:68]: Can you put in a note above (or as an inline comment) about `root`? In this example file is the first mention of it.

File: docs/src/_tutorials/logging.md [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+    propogate: no
+  urllib3: # for example hide all urllib3 debug logs
+    level: WARNING
+    handlers: [console, meltano_log]
+    propogate: no
+```
+
+## Local development config example
+
+While working with meltano locally it's sometimes nice to have more terse logging on the console, but still have DEBUG
```
Comment (tayloramurphy) [inline docs/src/_tutorials/logging.md:86]:
```suggestion
While working with Meltano locally it's sometimes nice to have more terse logging on the console, but still have DEBUG
```

File: docs/src/_tutorials/logging.md [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+
+root:
+  level: INFO
+  propagate: yes
+  handlers: [console]
+```
+
+## Datadog logging config
+
+For Datadog logging you have a couple of options. The easiest approach maybe to simply log to a file in json format and
```
Comment (tayloramurphy) [inline docs/src/_tutorials/logging.md:184]:
```suggestion
You have a couple options for structuring logs for Datadog. The easiest approach may be to log to a file in JSON format and
```

File: docs/src/_tutorials/logging.md [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+level info written to a log file behind scenes incase you need to debug something. To accomplish that, you can use a
+file like:
+
+```yaml
+version: 1
+disable_existing_loggers: false
+
+formatters:
+  structured_colored:
+    (): meltano.core.logging.console_log_formatter
```
Comment (tayloramurphy) [inline docs/src/_tutorials/logging.md:96]: How does a user know the world of available options here?

File: docs/src/_tutorials/logging.md [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+- `name` or `source` - Where the log message originated e.g. `tap-gitlab` if the log message originated from a Tap.
+- `stdio` - When the log message originated from a plugin, this field indicates whether the log message originated from stdout or stderr. Allowing you to filter out standard singer events for example.
+- `cmd_type` - The type of command that the log message originated from.
+- `state_id` - The associated state id.
+- `success` - Whether something succeeded or failed.
+- `error` - Where possible indicates the error type if one occurred.
+
+## Tips and tricks
+
+Use `[jq](https://stedolan.github.io/jq/)` to filter the output of json formatted meltano logs to only show the lines you're interested in.
```
Comment (tayloramurphy) [inline docs/src/_tutorials/logging.md:280]:
```suggestion
Use [jq](https://stedolan.github.io/jq/) to filter the output of JSON formatted Meltano logs to only show the lines you're interested in.
```

**Review by tayloramurphy — CHANGES_REQUESTED** [review]
Love this @pandemicsyn ! Just a few questions and suggestions.

**Comment by aaronsteers** [comment]
@pandemicsyn 

> any specific recipe we should include for CloudWatch alongside the Datadog and GCP ones or does the generic version cover it ?

I think the generic covers it! 👍 

Thanks for creating this. I think it will be a big help.

**Review by aaronsteers — APPROVED** [review]
Nothing blocking from my side. I like the comments/suggestions from @tayloramurphy, and I think a demo and walkthrough/talkthrough of these options could be super helpful sometime in the future.

**Review by edgarrmondragon — APPROVED** [review]
🎉 This looks great!

I wonder if this would feel more at home in `/concepts` or `/guide`. This document doesn't feel directed towards a learning experience or the user developing confidence with a tool or concept, in contrast to something like `Create a Custom Extractor`.

This all from what I grasped at the Diátaxis workshop:

> <p>A tutorial serves the needs of the user who is at study. Its obligation is <em>to provide a successful learning
> experience</em>. A how-to guide serves the needs of the user who is at work. Its obligation is <em>to help the user accomplish
> a task</em>. These are completely different needs and obligations, and they are why the distinction between tutorials and
> how-guides matters: tutorials are <strong>learning-oriented</strong>, and how-to guides are <strong>task-oriented</strong>.</p>

https://diataxis.fr/tutorials-how-to/

File: docs/src/_tutorials/logging.md (hunk 4 above)
Comment (pandemicsyn) [reply docs/src/_tutorials/logging.md:184]: This is less "structuring logs for Datadog" and more "configuring logs for Datadog". Mind if I roll with that phrasing ?

**Comment by pandemicsyn** [comment]
> 🎉 This looks great!
> 
> I wonder if this would feel more at home in `/concepts` or `/guide`. 

I'd be ok with any of those. Guide definitely makes a lot of sense.

File: docs/src/_tutorials/logging.md (hunk 5 above)
Comment (pandemicsyn) [reply docs/src/_tutorials/logging.md:96]: Doh good point, will update to document our current options.

File: docs/src/_tutorials/logging.md (hunk 2 above)
Comment (pandemicsyn) [reply docs/src/_tutorials/logging.md:68]: Good catch. My brain immediately assumed "folks will know with the root logger is for" when that's definitely not the case.

Comment (pandemicsyn) [reply docs/src/_tutorials/logging.md:72]: Added to description of all sections.

File: docs/src/_tutorials/logging.md (hunk 5 above)
Comment (pandemicsyn) [reply docs/src/_tutorials/logging.md:102]: updated. pointed at the src file that holds our formatters and listed what they currently are.

**Review by tayloramurphy — APPROVED** [review]

**Review by tayloramurphy — APPROVED** [review]

**Pull request merged**